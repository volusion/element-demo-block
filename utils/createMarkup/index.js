function createMarkup(htmlContentString = '') {
    const errorMessage =
        'Argument passed to "createMarkup()" function must be a string.';

    if (typeof htmlContentString !== 'string') throw new Error(errorMessage);
    return { __html: htmlContentString };
}

export { createMarkup };
