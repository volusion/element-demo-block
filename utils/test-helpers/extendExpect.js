expect.extend({
    toHaveContainerClassses(received) {
        const containerClasses = ['w-100', 'mw9', 'center'];
        const allClassesPresent = containerClasses.every(className =>
            received.hasClass(className)
        );
        if (allClassesPresent) {
            return {
                message: () =>
                    `Expected the block not to use the container classes: "w-100", "mw9", and "center".`,
                pass: true
            };
        } else {
            return {
                message: () =>
                    `Expected the block to be using the the container classes: "w-100", "mw9", and "center".`,
                pass: false
            };
        }
    }
});
export { expect };
