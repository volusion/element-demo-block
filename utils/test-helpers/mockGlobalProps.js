const mockGlobalProps = {
    color: {
        background: '#fff',
        primary: '#333',
        link: '#777',
        text: '#333'
    },
    globalComponents: {
        volComponentButton: {
            buttonStyles: {
                textColor: '#fff',
                textTransform: 'uppercase',
                backgroundColor: '#343538',
                borderColor: '#343538',
                hoverTextColor: '#fff',
                hoverBackgroundColor: '#555',
                hoverBorderColor: '#555',
                borderThickness: 'none', // basic, thin, thick, none
                size: 'large', // small, medium, large, 'block'
                rounded: 'none', // none, small, medium, large, pill
                growOnHover: false
            }
        }
    },
    typography: {
        baseFontSize: '16px',
        fontFamily:
            "-apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'",
        headingFontFamily:
            "-apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'",
        headingWeight: 700
    }
};
export { mockGlobalProps };
