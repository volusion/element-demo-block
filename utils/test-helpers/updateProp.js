export function updateProp(props = {}, propName, value) {
    expect(props).toHaveProperty(propName);
    props[propName] = value;
}
