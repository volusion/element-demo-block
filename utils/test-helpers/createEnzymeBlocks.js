import React from 'react';
import { Button, LinkButton } from '@volusion/element-components';
import { ElementPropTypes } from '@volusion/element-proptypes';
import { StyleSheet, css } from 'aphrodite';
import { shallow } from 'enzyme';
import { mockGlobalProps } from './mockGlobalProps';

function enzymeBlock(factory, blockConfig = {}, creationMethod = shallow) {
    if (typeof factory !== 'function')
        throw new Error(
            'First argument for "enzymeBlock()" must be a factory function.'
        );
    if (typeof blockConfig !== 'object')
        throw new Error(
            'Second argument for "enzymeBlock()" must be a block props object.'
        );
    const Button = () => <div>Mock Button</div>;
    const LinkButton = () => <div>Mock Button</div>;
    const mockComponents = { Button, LinkButton };
    const ReactBlock = factory(React, mockComponents);
    function processCss() {
        const arrayOfArguments = [...arguments];
        return arrayOfArguments
            .map(arg => {
                if (arg === undefined) return 'NO_MOCK_CLASS_PROVIDED';
                if (typeof arg === 'string') return arg;
                return css(arg);
            })
            .join(' ');
    }

    const enzymeBlockConfig = {
        ...blockConfig,
        StyleSheet,
        css: processCss
    };
    return creationMethod(<ReactBlock {...enzymeBlockConfig} />);
}

function createFactoryEnzymeBlock(
    factory,
    blockConfig = {},
    creationMethod = shallow
) {
    if (typeof factory !== 'function')
        throw new Error(
            'First argument for "createFactoryEnzymeBlock()" must be a factory function.'
        );
    if (typeof blockConfig !== 'object')
        throw new Error(
            'Second argument for "createFactoryEnzymeBlock()" must be a block props object.'
        );
    const Components = { Button, LinkButton };
    const utils = {};
    const globalConfig = { ...mockGlobalProps };
    const ReactBlock = factory(
        { React, ElementPropTypes, Components },
        utils,
        { StyleSheet, css },
        globalConfig,
        blockConfig
    ).block;
    const enzymeBlockConfig = { ...blockConfig, StyleSheet, css };
    return creationMethod(<ReactBlock {...enzymeBlockConfig} />);
}

export { createFactoryEnzymeBlock, enzymeBlock };
