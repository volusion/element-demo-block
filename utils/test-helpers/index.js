import { enzymeBlock, createFactoryEnzymeBlock } from './createEnzymeBlocks';
import { expect } from './extendExpect';
import { mockGlobalProps } from './mockGlobalProps';
import { updateProp } from './updateProp';

export {
    enzymeBlock,
    createFactoryEnzymeBlock,
    expect,
    mockGlobalProps,
    updateProp
};
