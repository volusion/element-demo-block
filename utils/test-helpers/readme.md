## Test Helper Functions

-   Isolates commonly used test helpers in a single global location.

### Import statment for all helpers

```javascript
import {
    enzymeBlock,
    createFactoryEnzymeBlock,
    expect,
    mockGlobalProps,
    updateProp
} from "../../utils/test-helpers";
```

## How each one works:

### enzymeBlock

-   Allows you to easily create an enzyme block of your block/component's factory function.
-   Faster than running the whole block through createFactory, and easier to parse the returned data.
-   Should always be utilized over `createFactoryEnzymeBlock` if possible.

```javascript
// Usage
const shallowBlock = enzymeBlock(factoryFn, props);
const mountedBlock = enzymeBlock(factoryFn, props, mount);
const renderedBlock = enzymeBlock(factoryFn, props, render);

// Example
import { enzymeBlock } from "../../utils/test-helpers";
import { factory as featuredContentFactory } from "...";

const props = {...};

describe("The Featured Content Block", () => {
    it("should have an Image component", () => {
        const shallowBlock = enzymeBlock(featuredContentFactory, props);
        expect(shallowBlock.exists("Image")).toBeTruthy();
    });
});
```

### createFactoryEnzymeBlock

-   Creates an enzyme element from the block after being passed through the createFactory function.
-   Useful when testing items that are handled as part of the createFactory function.

```javascript
// Usage
const shallowBlock = createFactoryEnzymeBlock(factoryFn, props);
const mountedBlock = createFactoryEnzymeBlock(factoryFn, props, mount);
const renderedBlock = createFactoryEnzymeBlock(factoryFn, props, render);

// Example
import { createFactoryEnzymeBlock } from "../../utils/test-helpers";
import { factory as featuredContentFactory } from "...";

const props = {...};

describe("The Featured Content Block", () => {
    it("should should use the default props if none are provided", () => {
        const shallowDefaultsBlock = createFactoryEnzymeBlock(featuredContentFactory, props);
        const shallowNoPropsBlock = createFactoryEnzymeBlock(featuredContentFactory, {});
        expect(shallowDefaultsBlock).toEqual(shallowNoPropsBlock);
    });
});
```

### expect

-   Extends expect to add a custom test function to check for container classes which every stand alone block should have.

```javascript
// Example
import { expect } from "../../utils/test-helpers";
expect(enzymeBlockNode).toHaveContainerClasses();
expect(enzymeBlockNode).not.toHaveContainerClasses();
```

### mockGlobalProps

-   Allows you to pass an object of the default global props into your tests.

```javascript
// Example
import { mockGlobalProps } from "../../utils/test-helpers";
import {styles} from '...';

const props = {...};

expect(styles(mockGlobalProps, props).backgroundColor).toEqual('...')
```

### updateProp

-   Validates the prop before updating it in your tests.
-   Prevents creating a new prop due to a typo or change in prop names.
-   If the prop you're trying to update does not exist, it throws an error in the tests.

```javascript
// Usage
updateProp(propsObj, "prop name as string", "value of prop");

// Example
import { updateProp } from "../../utils/test-helpers";
const props = { fullWidthSection: false };
updateProp(props, "fullWidthSection", true); // works
updateProp(props, "fullwidthSection", true); // throws an error because the `w` is the wrong case
```
