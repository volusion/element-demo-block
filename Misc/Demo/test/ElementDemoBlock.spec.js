import { ElementDemoBlockFactory } from '../src/Block';
import { defaultConfig } from '../src/Block/configs';
import { enzymeBlock } from '../../../utils/test-helpers';
import { mockClasses as classes } from './utils/mockClasses';

let props;

describe('The ElementDemoBlock block', () => {
    beforeEach(() => {
        props = { ...defaultConfig, classes };
    });
    it('should match the shallow snapshot', () => {
        const shallowWrapper = enzymeBlock(ElementDemoBlockFactory, props);
        expect(shallowWrapper).toMatchSnapshot();
    });
});
