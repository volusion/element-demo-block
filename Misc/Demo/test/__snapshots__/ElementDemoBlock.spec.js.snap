// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ElementDemoBlock block should match the shallow snapshot 1`] = `
<div
  className="pa3"
>
  <h1>
    I am heading text
  </h1>
  <ul>
    <li
      key="0"
    >
      I am the first item in the list.
    </li>
    <li
      key="1"
    >
      I am the second item in the list.
    </li>
  </ul>
  <OptionDetails
    selection="Option 1"
  />
  <img
    alt="Demo picture element"
    src="https://res.cloudinary.com/dyx4yhvoq/image/upload/h_350,w_350,c_pad/v1461691251/571a44b587179d0930f2c7d8/r1qpfyry8fz35evain24.jpg"
  />
  <div
    className="MOCK_CLASS_propsColor"
  >
    I get my color from "props.textColor"
  </div>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<p>I am a paragraph in an HTML string.</p>",
      }
    }
  />
</div>
`;
